### ISSUE TITLE
tracy_concurrentqueue.h warning with MSVC /permissive-

### ISSUE THREAD

**Issue by dubois** [wolfpld/tracy#403]
Verified with latest everything: Tracy v0.8.1, VS 2022 17.2.3, toolchain v143.
Compiler options: /std:c++17 or /std:c++20, and /permissive-

The tldr is that MSVC will still emit 4554 unless the pragmas surround the entire template like so:
```
#ifdef _MSC_VER
#pragma warning(push)
#pragma warning(disable: 4554)
#endif
	template<typename T>
	static inline bool circular_less_than(T a, T b)
	{
		static_assert(std::is_integral<T>::value && !std::numeric_limits<T>::is_signed, "circular_less_than is intended to be used only with unsigned integer types");
		return static_cast<T>(a - b) > static_cast<T>(static_cast<T>(1) << static_cast<T>(sizeof(T) * CHAR_BIT - 1));
	}
#ifdef _MSC_VER
#pragma warning(pop)
#endif
```

The fuller story is that we're starting the process of getting ready for /std:c++20 , and one of those steps is to start building in conformance mode (/permissive-). 

MSVC conformance mode changes the timing of its template instantiations, so it's likely that the above warning is a MSVC bug. Further evidence for it being a MSVC bug is that I _think_ I remember this working with earlier revisions of the v143 toolchain

Or, just as likely, MS might consider this new pragma warning behavior "working as intended", and the behavior changed because they were fixing conformance issues in v143. :shrug:

**Comment by wolfpld** [comment]
95808979
